src/chk/chk_rpc.c

```diff
@@ -0,0 +1,778 @@
+
+out:
+	D_CDEBUG(rc != 0, DLOG_ERR, DLOG_INFO,
+		 "Rank %u take action for DAOS check with gen "DF_X64", seq "DF_X64": "DF_RC"\n",
+		 rank, gen, seq, DP_RC(rc));
+
+	return rc;
+}
+
+int chk_report_remote(d_rank_t leader, uint64_t gen, uint32_t cla, uint32_t act, int32_t result,
+		      d_rank_t rank, uint32_t target, uuid_t *pool, uuid_t *cont,
+		      daos_unit_oid_t *obj, daos_key_t *dkey, daos_key_t *akey, char *msg,
+		      uint32_t option_nr, uint32_t *options, uint32_t detail_nr, d_sg_list_t *details,
+		      uint64_t *seq)
+{
+	crt_rpc_t		*req;
+	struct chk_report_in	*cri;
+	struct chk_report_out	*cro;
+	int			 rc;
+
+	rc = chk_sg_rpc_prepare(leader, CHK_REPORT, &req);
+	if (rc != 0)
+		goto out;
+
+	cri = crt_req_get(req);
+	cri->cri_gen = gen;
+	cri->cri_ics_class = cla;
+	cri->cri_ics_action = act;
+	cri->cri_ics_result = result;
+	cri->cri_rank = rank;
+	cri->cri_target = target;
+
+	if (pool != NULL)
+		uuid_copy(cri->cri_pool, *pool);
+	else
+		memset(cri->cri_pool, 0, sizeof(uuid_t));
+
+	if (cont != NULL)
+		uuid_copy(cri->cri_cont, *cont);
+	else
+		memset(cri->cri_cont, 0, sizeof(uuid_t));
+
+	if (obj != NULL)
+		cri->cri_obj = *obj;
+	else
+		cri->cri_obj = chk_dummy_obj;
+
+	if (dkey != NULL)
+		cri->cri_dkey = *dkey;
+	else
+		cri->cri_dkey = chk_dummy_key;
+
+	if (akey != NULL)
+		cri->cri_akey = *akey;
+	else
+		cri->cri_akey = chk_dummy_key;
+
+	cri->cri_msg = msg;
+	cri->cri_options.ca_count = option_nr;
+	cri->cri_options.ca_arrays = options;
+	cri->cri_details.ca_count = detail_nr;
+	cri->cri_details.ca_arrays = details;
+
+	rc = dss_rpc_send(req);
+	if (rc != 0)
+		goto out;
+
+	cro = crt_reply_get(req);
+	rc = cro->cro_status;
+	*seq = cro->cro_seq;
+
+out:
+	D_CDEBUG(rc != 0, DLOG_ERR, DLOG_INFO,
+		 "Rank %u report DAOS check to leader %u, gen "DF_X64", class %u, action %u, "
+		 "result %d, obj "DF_UOID", dkey "DF_KEY", akey "DF_KEY", msg %s, got seq "
+		 DF_X64": "DF_RC"\n", rank, leader, gen, cla, act, result,
+		 DP_UOID(obj != NULL ? *obj : chk_dummy_obj),
+		 DP_KEY(dkey != NULL ? dkey : &chk_dummy_key),
+		 DP_KEY(akey != NULL ? akey : &chk_dummy_key), msg, *seq, DP_RC(rc));
+
+	return rc;
+}
+
+int
+chk_rejoin_remote(d_rank_t leader, uint64_t gen, d_rank_t rank, uint32_t phase)
+{
+	crt_rpc_t		*req;
+	struct chk_rejoin_in	*cri;
+	struct chk_rejoin_out	*cro;
+	int			 rc;
+
+	rc = chk_sg_rpc_prepare(leader, CHK_REJOIN, &req);
+	if (rc != 0)
+		goto out;
+
+	cri = crt_req_get(req);
+	cri->cri_gen = gen;
+	cri->cri_rank = rank;
+	cri->cri_phase = phase;
+
+	rc = dss_rpc_send(req);
+	if (rc != 0)
+		goto out;
+
+	cro = crt_reply_get(req);
+	rc = cro->cro_status;
+
+out:
+	D_CDEBUG(rc != 0, DLOG_ERR, DLOG_INFO,
+		 "Rank %u rejoin DAOS check with leader %u, gen "DF_X64": "DF_RC"\n",
+		 rank, leader, gen, DP_RC(rc));
+
+	return rc;
+}
+
+static int
+crt_proc_daos_unit_oid_t(crt_proc_t proc, crt_proc_op_t proc_op, daos_unit_oid_t *p)
+{
+	return crt_proc_memcpy(proc, proc_op, p, sizeof(*p));
+}
```

Comment:
just fyi, if define as CRT_RAW (rather than CRT_VAR), may need not provider one special proc func to memcpy, as RAW will just memcpy to proc it. same for other flat mode struct filed, just like "(struct dtx_id)>       (orw_dti)>      >       CRT_RAW)"

((daos_unit_oid_t)	(cri_obj)		CRT_RAW)

Reply:
sure.